src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -251,69 +226,134 @@ protected virtual string GenerateUniqueId()
                                 .WriteStartMethodInvocation(_tagHelperContext.FormatInvalidIndexerAssignmentMethodName)
                                 .WriteStringLiteral(attributeName)
                                 .WriteParameterSeparator()
-                                .WriteStringLiteral(tagHelperTypeName)
+                                .WriteStringLiteral(tagHelperDescriptor.TypeName)
                                 .WriteParameterSeparator()
                                 .WriteStringLiteral(attributeDescriptor.PropertyName)
                                 .WriteEndMethodInvocation(endLine: false)   // End of method call
                                 .WriteEndMethodInvocation(endLine: true);   // End of new expression / throw statement
                         }
                     }
-
-                    var dictionaryKey = attributeName.Substring(attributeDescriptor.Name.Length);
-                    valueAccessor = string.Format(
-                        CultureInfo.InvariantCulture,
-                        "{0}.{1}[\"{2}\"]",
-                        tagHelperVariableName,
-                        attributeDescriptor.PropertyName,
-                        dictionaryKey);
                 }
+            }
+        }
+
+        private void RenderAttributes(
+            IList<KeyValuePair<string, Chunk>> chunkAttributes,
+            IEnumerable<TagHelperDescriptor> tagHelperDescriptors)
+        {
+            // First attribute wins, even if there are duplicates.
+            var renderedBoundAttributeNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
-                // If we haven't recorded this attribute value before then we need to record its value.
-                var attributeValueRecorded = htmlAttributeValues.ContainsKey(attributeName);
-                if (!attributeValueRecorded)
+            foreach (var attribute in chunkAttributes)
```

Comment:
Restore `// Go through the HTML attributes in source order, assigning to properties or indexers as we go.` comment, adding the possibility of adding HTML attribute to the execution context.